iree/compiler/Dialect/HAL/Target/WebGPU/WebGPUTarget.cpp

```diff
@@ -0,0 +1,270 @@
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::Codegen::IREECodegenDialect, spirv::SPIRVDialect,
+                    gpu::GPUDialect>();
+  }
+
+  IREE::HAL::DeviceTargetAttr getDefaultDeviceTarget(
+      MLIRContext *context) const override {
+    Builder b(context);
+    SmallVector<NamedAttribute> configItems;
+
+    configItems.emplace_back(b.getIdentifier("executable_targets"),
+                             getExecutableTargets(context));
+
+    auto configAttr = b.getDictionaryAttr(configItems);
+    return IREE::HAL::DeviceTargetAttr::get(
+        context, b.getStringAttr(deviceID()), configAttr);
+  }
+
+  void buildTranslationPassPipeline(OpPassManager &passManager) override {
+    buildSPIRVCodegenPassPipeline(passManager);
+    // TODO(scotttodd): additional passes for WebGPU/WGSL
+    //                  (here or during serialization?)
+  }
+
+  LogicalResult serializeExecutable(IREE::HAL::ExecutableVariantOp variantOp,
+                                    OpBuilder &executableBuilder) override {
+    ModuleOp innerModuleOp = variantOp.getInnerModule();
+    auto spirvModuleOps = innerModuleOp.getOps<spirv::ModuleOp>();
+    if (!llvm::hasSingleElement(spirvModuleOps)) {
+      return variantOp.emitError()
+             << "should only contain exactly one spv.module op";
+    }
+    auto spvModuleOp = *spirvModuleOps.begin();
+
+    // The schema expects each shader module to have entry points named "dN",
+    // where N is the entry point ordinal.
+    // For each executable entry point op, rename the entry point symbol using
+    // that convention and keep track of the mapping between entry point
+    // ordinals to which shader module they reference.
+    auto entryPointOps = llvm::to_vector<4>(
+        variantOp.getOps<IREE::HAL::ExecutableEntryPointOp>());
+    llvm::SmallVector<uint32_t, 4> entryPointOrdinals(entryPointOps.size());
+    SymbolTableCollection symbolTable;
+    SymbolUserMap symbolUsers(symbolTable, variantOp);
+    for (auto entryPointOp : entryPointOps) {
+      auto entryPointFunc = dyn_cast<spirv::FuncOp>(
+          SymbolTable::lookupSymbolIn(spvModuleOp, entryPointOp.sym_name()));
+
+      std::string symbolName = llvm::formatv("d{0}", entryPointOp.ordinal());
+      mlir::StringAttr nameAttr =
+          mlir::StringAttr::get(variantOp->getContext(), symbolName);
+
+      symbolUsers.replaceAllUsesWith(entryPointFunc, nameAttr);
+      entryPointOp.setName(symbolName);  // Same symbol reference? Not in table?
+      SymbolTable::setSymbolName(entryPointFunc, symbolName);
+
+      // We only have one shader module right now, so all point to index 0.
+      // TODO(#7824): Support multiple shader modules per executable
+      entryPointOrdinals[entryPointOp.ordinal().getZExtValue()] = 0;
+    }
+
+    // Serialize the spirv::ModuleOp into binary format.
+    SmallVector<uint32_t, 256> spvBinary;
```

Comment:
nothing will ever be under 256 bytes and will incur the heap allocation so you can shrink this to save some stack space

Reply:
some divergence between the similar compiler targets:

https://github.com/google/iree/blob/4af202b14f4dabb6ac116836eb05354e8ba1de26/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp#L291-L294

https://github.com/google/iree/blob/4af202b14f4dabb6ac116836eb05354e8ba1de26/iree/compiler/Dialect/HAL/Target/MetalSPIRV/MetalSPIRVTarget.cpp#L82-L84